src/contribute/runner-guide.md

```diff
@@ -0,0 +1,1103 @@
+
+#### Reading from a BoundedSource
+
+A `BoundedSource` is a source of data that you know is finite, such as a static
+collection of log files, or a database table. The capabilities are:
+
+ * `split(int)` - your runner should call this to get desired initial parallelism (but you can often steal work later)
+ * `getEstimatedSizeBytes(...)` - self explanatory
+ * `createReader(...)` - call this to start reading elements; it is an enhanced iterator, with also:
+ * timestamps to associate with each element read
+ * `splitAtFraction` for dynamic splitting to enable work stealing, and other
+   methods to support it - see the [Beam blog post on dynamic work
+   rebalancing](https://beam.apache.org/blog/2016/05/18/splitAtFraction-method.html)
+
+The `BoundedSource` does not report a watermark currently. Most of the time, reading
+from a bounded source can be parallelized in ways that result in utterly out-of-order
+data, so a watermark is not terribly useful.
+Thus the watermark for the output `PCollection` from a bounded read should
+remain at the minimum timestamp throughout reading (otherwise data might get
+dropped) and advance to the maximum timestamp when all data is exhausted.
+
+### Implementing the Flatten primitive
+
+This one is easy - take as input a finite set of `PCollections` and outputs their
+bag union, keeping windows intact.
+
+For this operation to make sense, it is the SDK's responsibility to make sure
+the windowing strategies are compatible.
+
+Also note that there is no requirement that the coders for all the `PCollections`
+be the same. If your runner wants to require that (to avoid tedious
+re-encoding) you have to enforce it yourself. Or you could just implement the
+fast path as an optimization.
+
+### Special mention: the Combine composite
+
+A composite transform that is almost always treated specially by a runner is
+`Combine` (per key), which applies an associative and commutative operator to
+the elements of a `PCollection`. This composite is not a primitive. It is
+implemented in terms of `ParDo` and `GroupByKey`, so your runner will work
+without treating it - but it does carry additional information that you
+probably want to use for optimizations: the associative-commutative operator,
+known as a `CombineFn`.
+
+## Working with pipelines
+
+When you receive a pipeline from a user, you will need to translate it. This is
+a tour of the APIs that you'll use to do it.  
+
+### Traversing a pipeline
+
+Something you will certain do is to traverse a pipeline, probably to
+translate it into primitives for your engine. The general pattern is
+to write a visitor that builds a job specification as it walks the
+graph of `PTransforms`.
+
+The entry point for this in Java is
+[`Pipeline.traverseTopologically`](https://beam.apache.org/documentation/sdks/javadoc/2.0.0/org/apache/beam/sdk/Pipeline.html#traverseTopologically-org.apache.beam.sdk.Pipeline.PipelineVisitor-)
```

Comment:
That seems reasonable...

I just tried it, but it seems that we redirect `https://beam.apache.org/documentation/sdks/javadoc/` to `https://beam.apache.org/documentation/sdks/javadoc/2.0.0` at the root but it doesn't work for deeper links.

I also tried `latest` which is pretty common, but that is also a 404.

So if we set something like that up, I would definitely use it.